casbin/casbin.go

```diff
@@ -91,44 +103,42 @@ func Middleware(ce *casbin.Enforcer) echo.MiddlewareFunc {
 // MiddlewareWithConfig returns a CasbinAuth middleware with config.
 // See `Middleware()`.
 func MiddlewareWithConfig(config Config) echo.MiddlewareFunc {
-	// Defaults
+	if config.Enforcer == nil && config.EnforceHandler == nil {
+		panic("one of casbin middleware Enforcer or EnforceHandler fields must be set")
+	}
 	if config.Skipper == nil {
 		config.Skipper = DefaultConfig.Skipper
 	}
+	if config.UserGetter == nil {
+		config.UserGetter = DefaultConfig.UserGetter
+	}
+	if config.ErrorHandler == nil {
+		config.ErrorHandler = DefaultConfig.ErrorHandler
+	}
+	if config.EnforceHandler == nil {
+		config.EnforceHandler = func(c echo.Context, user string) (bool, error) {
+			return config.Enforcer.Enforce(user, c.Request().URL.Path, c.Request().Method)
+		}
+	}
 
 	return func(next echo.HandlerFunc) echo.HandlerFunc {
 		return func(c echo.Context) error {
 			if config.Skipper(c) {
 				return next(c)
 			}
 
-			if pass, err := config.CheckPermission(c); err == nil && pass {
-				return next(c)
-			} else if err != nil {
-				return echo.NewHTTPError(http.StatusInternalServerError, err.Error())
+			user, err := config.UserGetter(c)
+			if err != nil {
+				return config.ErrorHandler(c, err, http.StatusForbidden)
 			}
-
-			return echo.ErrForbidden
+			pass, err := config.EnforceHandler(c, user)
+			if err != nil {
+				return config.ErrorHandler(c, err, http.StatusInternalServerError)
+			}
+			if !pass {
+				return config.ErrorHandler(c, errors.New("enforce did not pass"), http.StatusForbidden)
```

Comment:
Can you make the status code configurable? For example, one might use 402 instead of 403.

Reply:
this can be done inside/with errorhandler.